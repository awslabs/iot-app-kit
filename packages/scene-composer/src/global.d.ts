declare module '*.hdr';
declare module '*.png';
