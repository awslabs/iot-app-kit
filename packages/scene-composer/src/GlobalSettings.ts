import { COMPOSER_FEATURES, FeatureConfig, GetSceneObjectFunction } from './interfaces';
import { IMetricRecorder } from './interfaces/metricRecorder';

export interface DracoDecoderConfig {
  enable: boolean;
  path?: string;
}

const globalSettings: {
  debugMode: boolean;
  dracoDecoder: DracoDecoderConfig;
  locale: string;
  cdnPath: string | undefined;
  shouldEnableDataBindingTemplate: boolean;
  metricRecorder?: IMetricRecorder;
  featureConfig: FeatureConfig;
  getSceneObjectFunction: GetSceneObjectFunction | undefined;
} = {
  debugMode: false,
  dracoDecoder: { enable: true },
  locale: 'en-US',
  cdnPath: undefined,
  shouldEnableDataBindingTemplate: false,
  metricRecorder: undefined,
  // default to disable all features
  featureConfig: {},
  getSceneObjectFunction: undefined,
};

const changeSubscribers = [] as Function[];

const notifySubscribers = () => {
  changeSubscribers.forEach((sub) => sub());
};

export const setDebugMode = () => {
  globalSettings.debugMode = true;
  notifySubscribers();
};

export const setDracoDecoder = (dracoDecoder: DracoDecoderConfig) => {
  globalSettings.dracoDecoder = dracoDecoder;
  notifySubscribers();
};
export const setLocale = (locale: string) => {
  globalSettings.locale = locale;
  notifySubscribers();
};

export const setCdnPath = (path: string | undefined) => {
  globalSettings.cdnPath = path;
  notifySubscribers();
};

// Before we have feature flag, we use global settings to control this new feature
export const enableDataBindingTemplate = () => {
  globalSettings.shouldEnableDataBindingTemplate = true;
  notifySubscribers();
};

export const setMetricRecorder = (metricRecorder: IMetricRecorder) => {
  globalSettings.metricRecorder = metricRecorder;
  notifySubscribers();
};

export const setFeatureConfig = (featureConfig: Partial<Record<COMPOSER_FEATURES, boolean>>) => {
  globalSettings.featureConfig = featureConfig;
  notifySubscribers();
};

export const setGetSceneObjectFunction = (getSceneObjectFunction: GetSceneObjectFunction) => {
  globalSettings.getSceneObjectFunction = getSceneObjectFunction;
  notifySubscribers();
};

export const getGlobalSettings = (): Readonly<typeof globalSettings> => {
  return globalSettings;
};

export const subscribe = (callback: Function) => {
  changeSubscribers.push(callback);
};

export const unsubscribe = (callback: Function) => {
  changeSubscribers.splice(changeSubscribers.indexOf(callback), 1);
};
