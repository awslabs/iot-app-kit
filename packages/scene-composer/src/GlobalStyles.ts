import { createGlobalStyle } from 'styled-components';

export const GlobalStyles = createGlobalStyle`
  .tm-side-panel-tabs {
    height: 100vh;
    max-width: 100vw;
    overflow: auto;
  }
  .tm-side-panel-tabs > div:first-of-type {
    padding: 0 !important; // This removes the padding of the AWSUI Tabs Component
  }`;
