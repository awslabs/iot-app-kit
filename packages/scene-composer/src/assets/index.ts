export * from './anchors';
export * from './viewpoints';
