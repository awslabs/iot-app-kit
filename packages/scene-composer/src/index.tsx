export * from './interfaces';

export * from './common/errors';

export * from './components/SceneComposerInternal';

export * from './SceneViewer';

export type { Scene } from './models/SceneModels';

// Augmentation related interfaces
export * from './augmentations/three';

export * from './common/GlobalSettings';

export * from './utils/dataFrameLabelUtils';

export * from './utils/dataBindingTemplateUtils';

export * from './common/constants';
