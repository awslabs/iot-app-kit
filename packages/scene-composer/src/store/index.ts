export * from './StoreOperations';
export * from './Store';
