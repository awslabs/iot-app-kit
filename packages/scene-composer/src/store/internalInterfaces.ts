import { IErrorDetails } from '../common/errors';
import {
  SelectionChangedEventCallback,
  IAnchorComponent,
  ICameraComponent,
  IColorOverlayComponent,
  ILightComponent,
  IModelRefComponent,
  IMotionIndicatorComponent,
  IRuleBasedMap,
  IRuleStatement,
  ISceneComponent,
  ISceneDocument,
  ISceneNode,
  ITransform,
  ITransformConstraint,
  IValueDataBinding,
  IValueDataBindingProvider,
  OperationMode,
  ShowAssetBrowserCallback,
  URIModifier,
  KnownComponentType,
  WidgetClickEventCallback,
  ISubModelRefComponent,
  IDataOverlayComponent,
} from '../interfaces';
import { MapControls as MapControlsImpl, OrbitControls as OrbitControlsImpl } from '../three/OrbitControls';

export type ISerializationErrorDetails = IErrorDetails;

export interface IDeserializationResult {
  document?: ISceneDocumentInternal;
  errors: ISerializationErrorDetails[];
}

/******************************************************************************
 * Editor Data Model
 ******************************************************************************/

export enum DisplayMessageCategory {
  Error = 'Error',
  Warning = 'Warning',
  Info = 'Info',
}

export type CameraControlImpl = MapControlsImpl | OrbitControlsImpl;
export type TweenValueObject = { x: number; y: number; z: number };

export type CursorStyle = 'move' | 'edit';

export interface IDisplayMessage {
  category: DisplayMessageCategory;
  messageText: string;
  params?: Record<string, string>;
}

export interface IEditorConfig {
  operationMode?: OperationMode;
  uriModifier?: URIModifier;
  valueDataBindingProvider?: IValueDataBindingProvider;
  showAssetBrowserCallback?: ShowAssetBrowserCallback;
  onWidgetClick?: WidgetClickEventCallback;
  onSelectionChanged?: SelectionChangedEventCallback;
}

/******************************************************************************
 * Document Model
 ******************************************************************************/

export type ITransformInternal = ITransform;
export type IRuleBasedMapInternal = IRuleBasedMap;
export type IRuleStatementInternal = IRuleStatement;

export interface ISceneNodeInternal extends ISceneNode {
  // below fields are not nullable internally
  ref: string;
  name: string;
  transform: ITransformInternal;
  transformConstraint: ITransformConstraint;
  components: ISceneComponentInternal[];
  childRefs: string[];
  properties: Record<string, any>;
}

export interface ISceneDocumentInternal extends ISceneDocument {
  unit: string; // unit is not nullable internally, default to 'meter'
  ruleMap: Record<string, IRuleBasedMapInternal>;
  nodeMap: Record<string, ISceneNodeInternal>;
  componentNodeMap: { [type in KnownComponentType | string]?: Record<string, string[]> }; // record of node.ref to component.ref[]
}

export interface ISceneComponentInternal extends ISceneComponent {
  // unique id for the component for internal reference
  ref: string;
}

export interface IDataBoundSceneComponentInternal extends ISceneComponentInternal {
  valueDataBinding?: IValueDataBinding;
}

/******************************************************************************
 * Components
 ******************************************************************************/

export type ILightComponentInternal = ISceneComponentInternal & ILightComponent;

export type ICameraComponentInternal = ISceneComponentInternal & ICameraComponent;

export type IModelRefComponentInternal = ISceneComponentInternal & IModelRefComponent;

export type ISubModelRefComponentInternal = ISceneComponentInternal & ISubModelRefComponent;

export type IAnchorComponentInternal = IDataBoundSceneComponentInternal & IAnchorComponent;

export type IColorOverlayComponentInternal = IDataBoundSceneComponentInternal & IColorOverlayComponent;

export type IMotionIndicatorComponentInternal = ISceneComponentInternal & IMotionIndicatorComponent;

export type IDataOverlayComponentInternal = ISceneComponentInternal & IDataOverlayComponent;

/******************************************************************************
 * Type magic...
 ******************************************************************************/

export const isISceneNodeInternal = (item: ISceneNodeInternal | undefined): item is ISceneNodeInternal => {
  return !!item;
};

export const isISceneComponentInternal = (
  item: ISceneComponentInternal | undefined,
): item is ISceneComponentInternal => {
  return !!item;
};
