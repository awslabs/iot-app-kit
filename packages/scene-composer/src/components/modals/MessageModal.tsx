import _ from 'lodash';
import { Box, Button, Header, Icon, SpaceBetween, StatusIndicator, TextContent } from '@awsui/components-react';
import React, { Fragment, useContext, useMemo } from 'react';
import { useIntl } from 'react-intl';

import { AWS_USER_GUIDE_DOC_URL } from '../../common/constants';
import { sceneComposerIdContext } from '../../common/sceneComposerIdContext';
import { useStore } from '../../store';
import { DisplayMessageCategory } from '../../store/internalInterfaces';
import { evalStringTemplate } from '../../utils/stringUtils';
import CenteredContainer from '../CenteredContainer';

function mapMessageCategoryToModalTitle(category: DisplayMessageCategory) {
  const intl = useIntl();
  switch (category) {
    case DisplayMessageCategory.Error:
      return intl.formatMessage({ defaultMessage: 'Error - unable to render the scene', description: 'Error message' });
    case DisplayMessageCategory.Warning:
      return intl.formatMessage({ defaultMessage: 'Warning', description: 'Warning message' });
    default:
      return intl.formatMessage({ defaultMessage: 'Notice', description: 'Notice message' });
  }
}

const MessageModal = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const messages = useStore(sceneComposerId)((state) => state.getMessages());
  const clearMessages = useStore(sceneComposerId)((state) => state.clearMessages);
  const intl = useIntl();

  const containsError = messages.find((m) => m.category === DisplayMessageCategory.Error);
  const containsWarning = messages.find((m) => m.category === DisplayMessageCategory.Warning);

  const title = containsError
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Error)
    : containsWarning
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Warning)
    : mapMessageCategoryToModalTitle(DisplayMessageCategory.Info);
  const allowDismiss = !containsError;
  const showUserGuideLink = containsError || containsWarning;

  const messageContent = useMemo(() => {
    const groupedMessages = _.groupBy(messages, (m) => m.category);
    return [DisplayMessageCategory.Error, DisplayMessageCategory.Warning, DisplayMessageCategory.Info]
      .map((category) => {
        if (category in groupedMessages) {
          const messages = groupedMessages[category];
          if (messages.length === 0) {
            return null;
          }
          return (
            <Fragment key={category}>
              {messages.map((m, index) => (
                <StatusIndicator key={index} type={category.toLocaleLowerCase() as any}>
                  {evalStringTemplate(m.messageText, m.params)}
                  {m.params?.cause ? ' Cause - ' + m.params.cause : ''}
                </StatusIndicator>
              ))}
            </Fragment>
          );
        } else {
          return null;
        }
      })
      .filter((m) => m !== null);
  }, [messages]);

  return (
    <CenteredContainer
      header={<Header variant='h2'>{title}</Header>}
      footer={
        allowDismiss && (
          <Box float='right' padding={{ bottom: 's' }}>
            <Button variant='primary' onClick={clearMessages}>
              Ok
            </Button>
          </Box>
        )
      }
    >
      <SpaceBetween size='s'>
        <Box>{messageContent}</Box>

        {showUserGuideLink && (
          <Box>
            <TextContent>
              <p>
                <a href={AWS_USER_GUIDE_DOC_URL}>
                  {' '}
                  {intl.formatMessage({ defaultMessage: 'Learn More', description: 'Hyperlink title' })}{' '}
                  <Icon name='external' size='normal' variant='link' />{' '}
                </a>
              </p>
            </TextContent>
          </Box>
        )}
      </SpaceBetween>
    </CenteredContainer>
  );
};

MessageModal.displayName = 'MessageModal';

export default MessageModal;
