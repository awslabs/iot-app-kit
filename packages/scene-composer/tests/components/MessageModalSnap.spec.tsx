import renderer from 'react-test-renderer';
import React from 'react';

import { useStore } from '../../src/store';
import MessageModal from '../../src/components/MessageModal';
import { DisplayMessageCategory } from '../../src/store/internalInterfaces';

import Mock = jest.Mock;

jest.mock('../../src/store', () => {
  const originalModule = jest.requireActual('../../src/store');
  return {
    ...originalModule,
    useStore: jest.fn(),
  };
});

describe('MessageModalSnap', () => {
  const clearMessages = jest.fn();
  const mockUseStore = useStore as Mock;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with info', () => {
    mockUseStore.mockReturnValueOnce(() => [{ category: DisplayMessageCategory.Info, messageText: 'Info Test' }]);
    mockUseStore.mockReturnValueOnce(() => clearMessages);

    const container = renderer.create(<MessageModal />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with warning', () => {
    mockUseStore.mockReturnValueOnce(() => [{ category: DisplayMessageCategory.Warning, messageText: 'Warning Test' }]);
    mockUseStore.mockReturnValueOnce(() => clearMessages);

    const container = renderer.create(<MessageModal />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with error', () => {
    mockUseStore.mockReturnValueOnce(() => [{ category: DisplayMessageCategory.Error, messageText: 'Error Test' }]);
    mockUseStore.mockReturnValueOnce(() => clearMessages);

    const container = renderer.create(<MessageModal />);
    expect(container).toMatchSnapshot();
  });
});
