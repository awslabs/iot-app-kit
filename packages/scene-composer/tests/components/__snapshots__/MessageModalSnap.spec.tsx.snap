// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageModalSnap should render correctly with error 1`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div
  className="c0"
  data-mocked="Container"
  footer={false}
>
  <div>
    <div
      data-mocked="Header"
      variant="h2"
    >
      Error - unable to render the scene
    </div>
  </div>
  <div
    data-mocked="SpaceBetween"
    size="s"
  >
    <div
      data-mocked="Box"
    >
      <div
        data-mocked="StatusIndicator"
        type="error"
      >
        Error Test
      </div>
    </div>
    <div
      data-mocked="Box"
    >
      <div
        data-mocked="TextContent"
      >
        <p>
          <a
            href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
          >
             
            Learn More
             
            <div
              data-mocked="Icon"
              name="external"
              size="normal"
              variant="link"
            />
             
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap should render correctly with info 1`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div
  className="c0"
  data-mocked="Container"
>
  <div>
    <div
      data-mocked="Header"
      variant="h2"
    >
      Notice
    </div>
  </div>
  <div>
    <div
      data-mocked="Box"
      float="right"
      padding="{\\"bottom\\":\\"s\\"}"
    >
      <div
        data-mocked="Button"
        onClick={[MockFunction]}
        variant="primary"
      >
        Ok
      </div>
    </div>
  </div>
  <div
    data-mocked="SpaceBetween"
    size="s"
  >
    <div
      data-mocked="Box"
    >
      <div
        data-mocked="StatusIndicator"
        type="info"
      >
        Info Test
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap should render correctly with warning 1`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div
  className="c0"
  data-mocked="Container"
>
  <div>
    <div
      data-mocked="Header"
      variant="h2"
    >
      Warning
    </div>
  </div>
  <div>
    <div
      data-mocked="Box"
      float="right"
      padding="{\\"bottom\\":\\"s\\"}"
    >
      <div
        data-mocked="Button"
        onClick={[MockFunction]}
        variant="primary"
      >
        Ok
      </div>
    </div>
  </div>
  <div
    data-mocked="SpaceBetween"
    size="s"
  >
    <div
      data-mocked="Box"
    >
      <div
        data-mocked="StatusIndicator"
        type="warning"
      >
        Warning Test
      </div>
    </div>
    <div
      data-mocked="Box"
    >
      <div
        data-mocked="TextContent"
      >
        <p>
          <a
            href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
          >
             
            Learn More
             
            <div
              data-mocked="Icon"
              name="external"
              size="normal"
              variant="link"
            />
             
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
